src/openrct2/windows/view_clipping.c

```diff
@@ -0,0 +1,285 @@
+	WIDX_CLIP_HEIGHT_DECREASE,
+	WIDX_CLIP_HEIGHT_SLIDER
+};
+
+typedef enum DISPLAY_TYPE {
+	DISPLAY_RAW = 0,
+	DISPLAY_UNITS = 1
+} clip_value_display_type;
+
+clip_value_display_type gClipHeightDisplayType = DISPLAY_UNITS;
+
+#pragma region Widgets
+
+#define WW 160
+#define WH 70
+
+rct_widget window_view_clipping_widgets[] = {
+	{ WWT_FRAME,		0,	0,		WW - 1,	0,	WH - 1, STR_NONE,				STR_NONE }, // panel / background
+	{ WWT_CAPTION,		0,	1,		WW - 2,	1,	14,     STR_VIEW_CLIPPING_TITLE,		STR_WINDOW_TITLE_TIP }, // title bar
+	{ WWT_CLOSEBOX,		0,	WW - 13,	WW - 3,	2,	13,     STR_CLOSE_X,				STR_CLOSE_WINDOW_TIP }, // close x button
+	{ WWT_CHECKBOX,		0,	11,		149,	19,	29,     STR_VIEW_CLIPPING_HEIGHT_ENABLE,	STR_VIEW_CLIPPING_HEIGHT_ENABLE_TIP }, // clip height enable/disable check box
+	{ WWT_SPINNER,		0,	90,		149,	34,	45,     STR_NONE,				STR_VIEW_CLIPPING_HEIGHT_VALUE_TOGGLE }, // clip height value
+	{ WWT_DROPDOWN_BUTTON,	0,	138,		148,	35,	39,     STR_NUMERIC_UP,				STR_NONE }, // clip height increase
+	{ WWT_DROPDOWN_BUTTON, 0,	138,		148,	40,	44,     STR_NUMERIC_DOWN,			STR_NONE }, // clip height decrease
+	{ WWT_SCROLL,          0,	11,		149,	49,	61,     SCROLL_HORIZONTAL,			STR_VIEW_CLIPPING_HEIGHT_SCROLL_TIP }, // clip height scrollbar
+	{ WIDGETS_END }
+};
+
+#pragma endregion
+
+#pragma region Events
+
+static void window_view_clipping_close_button(rct_window *w);
+static void window_view_clipping_mouseup(rct_window *w, int widgetIndex);
+static void window_view_clipping_update(rct_window *w);
+static void window_view_clipping_invalidate(rct_window *w);
+static void window_view_clipping_paint(rct_window *w, rct_drawpixelinfo *dpi);
+static void window_view_clipping_scrollgetsize(rct_window *w, int scrollIndex, int *width, int *height);
+
+static rct_window_event_list window_view_clipping_events = {
+	window_view_clipping_close_button,
+	window_view_clipping_mouseup,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	window_view_clipping_update,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	window_view_clipping_scrollgetsize,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	NULL,
+	window_view_clipping_invalidate,
+	window_view_clipping_paint,
+	NULL
+};
+
+#pragma endregion
+
+static void window_view_clipping_set_clipheight(rct_window *w, const uint8 clipheight)
+{
+	gClipHeight = clipheight;
+	rct_widget* widget = &window_view_clipping_widgets[WIDX_CLIP_HEIGHT_SLIDER];
+	const float clip_height_ratio = (float)gClipHeight / 255;
+	w->scrolls[0].h_left = (sint16)ceil(clip_height_ratio * (w->scrolls[0].h_right - ((widget->right - widget->left) - 1)));
+}
+
+void window_view_clipping_open()
+{
+	rct_window* window;
+
+	// Get the main viewport to set the view clipping flag.
+	rct_window *mainWindow = window_get_main();
+
+	// Check if window is already open
+	if (window_find_by_class(WC_VIEW_CLIPPING) != NULL) {
+		// If window is already open, toggle the view clipping on/off
+		if (mainWindow != NULL) {
```

Comment:
It seems redunant to me, the main window will always be opened. As you're getting the pointer to it in the same function (not storing it) it's also not affected by the window list being reordered.